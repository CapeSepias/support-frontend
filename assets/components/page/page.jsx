// @flow

// ----- Imports ----- //

import React, { type Node } from 'react';


// ----- Types ----- //

type PropTypes = {
  id: ?string,
  header: Node,
  footer: Node,
  children: Node,
  optChildren: ?Node
};


// ----- Component ----- //

export default function Page(props: PropTypes) {

  return (
    <div id={props.id} className="gu-content">
      {props.header}
      <main role="main" className="gu-content__main">
        {props.children}
      </main>
      {props.footer}
    </div>
  );

}


// ----- Default Props ----- //

Page.defaultProps = {
  id: null,
};
