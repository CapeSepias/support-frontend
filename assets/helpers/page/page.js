// @flow

// ----- Imports ----- //

import { createStore, combineReducers } from 'redux';
import 'ophan';

import * as ga from 'helpers/tracking/ga';
import * as abTest from 'helpers/abtest';
import * as logger from 'helpers/logger';
import * as googleTagManager from 'helpers/tracking/googleTagManager';
import { getCampaign, getAcquisition } from 'helpers/tracking/acquisitions';
import { detect } from 'helpers/internationalisation/country';

import type { Campaign, ReferrerAcquisitionData } from 'helpers/tracking/acquisitions';
import type { IsoCountry } from 'helpers/internationalisation/country';
import type { Participations } from 'helpers/abtest';
import type { Dimensions } from 'helpers/tracking/googleTagManager';
import { getQueryParams, getQueryParameter } from 'helpers/url';


import type { Action } from './pageActions';


// ----- Types ----- //

export type CommonState = {
  campaign: ?Campaign,
  referrerAcquisitionData: ReferrerAcquisitionData,
  otherQueryParams: Array<[string, string]>,
  country: IsoCountry,
  abParticipations: Participations,
};

export type PreloadedState = {
  campaign?: $PropertyType<CommonState, 'campaign'>,
  referrerAcquisitionData?: $PropertyType<CommonState, 'referrerAcquisitionData'>,
  country?: $PropertyType<CommonState, 'country'>,
  abParticipations?: $PropertyType<CommonState, 'abParticipations'>,
};


// ----- Functions ----- //

// Sets up GA and logging.
function analyticsInitialisation(participations: Participations): void {

  // Google analytics.
  ga.init();
  ga.setDimension('campaignCodeBusinessUnit', getQueryParameter('CMP_BUNIT'));
  ga.setDimension('campaignCodeTeam', getQueryParameter('CMP_TU'));
  ga.setDimension('experience', abTest.getVariantsAsString(participations));
  ga.trackPageview();

  const dimensions:Dimensions = {
    campaignCodeBusinessUnit: getQueryParameter('CMP_BUNIT') || undefined,
    campaignCodeTeam: getQueryParameter('CMP_TU') || undefined,
    experience: abTest.getVariantsAsString(participations),
  };

  googleTagManager.init();
  googleTagManager.pushDimensions(dimensions);

  // Logging.
  logger.init();

}

// Creates the initial state for the common reducer.
function buildInitialState(
  abParticipations: Participations,
  preloadedState: PreloadedState = {},
  country: IsoCountry,
): CommonState {

  const acquisition = getAcquisition();
  const otherQueryParams = getQueryParams(['REFPVID', 'INTCMP']);

  return Object.assign({}, {
    campaign: acquisition ? getCampaign(acquisition) : null,
    referrerAcquisitionData: acquisition,
    otherQueryParams,
    country,
    abParticipations,
  }, preloadedState);

}

// Sets up the common reducer with its initial state.
function createCommonReducer(
  initialState: CommonState): (CommonState, Action) => CommonState {

  function commonReducer(
    state: CommonState = initialState,
    action: Action): CommonState {

    switch (action.type) {

      case 'SET_COUNTRY':
        return Object.assign({}, state, { country: action.country });

      default:
        return state;

    }

  }

  return commonReducer;

}

// For pages that don't need Redux.
function statelessInit() {
  const country: IsoCountry = detect();
  const participations: Participations = abTest.init(country);
  analyticsInitialisation(participations);

}

// Initialises the page.
function init(
  pageReducer: Object,
  preloadedState?: PreloadedState,
  middleware: ?Function,
) {

  const country: IsoCountry = detect();
  const participations: Participations = abTest.init(country);
  analyticsInitialisation(participations);
  const initialState: CommonState = buildInitialState(participations, preloadedState, country);
  const commonReducer = createCommonReducer(initialState);

  return createStore(
    combineReducers({ page: pageReducer, common: commonReducer }),
    undefined,
    middleware,
  );

}


// ----- Exports ----- //

export {
  createCommonReducer,
  init,
  statelessInit,
};
