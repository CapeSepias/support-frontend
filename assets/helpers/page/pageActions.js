// @flow

// ----- Imports ----- //

import type { IsoCountry } from 'helpers/internationalisation/country';


// ----- Types ----- //

export type Action = { type: 'SET_COUNTRY', country: IsoCountry };


// ----- Action Creators ----- //

export function setCountry(country: IsoCountry): Action {
  return { type: 'SET_COUNTRY', country };
}
