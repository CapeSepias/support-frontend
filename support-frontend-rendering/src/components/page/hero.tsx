/** @jsx jsx */
import { jsx, SerializedStyles } from '@emotion/react';
import { Breakpoint } from '@guardian/src-foundations/dist/types/mq';
import React from 'react';
import {
    hero,
    heroRoundel,
    heroRoundelContainer,
    heroImage,
    roundelOffset,
    roundelNudgeDown,
    roundelNudgeUp,
    roundelHidingPoints,
} from './heroStyles';

// Options for moving the roundel position on mobile
type RoundelNudgeDirection = 'up' | 'down' | 'none';
type PropTypes = {
    image: React.ReactChild;
    children: React.ReactChild;
    cssOverrides?: string;
    roundelText?: React.ReactChild;
    roundelNudgeDirection?: RoundelNudgeDirection;
    hideRoundelBelow?: Breakpoint;
};

const roundelNudges: { [key in RoundelNudgeDirection]: SerializedStyles | string } = {
    up: roundelNudgeUp,
    down: roundelNudgeDown,
    none: '',
};

function Hero({
    children,
    image,
    cssOverrides,
    roundelText,
    roundelNudgeDirection = 'up',
    hideRoundelBelow,
}: PropTypes): React.ReactElement {
    const useOffset = roundelText && roundelNudgeDirection === 'up';
    const nudgeCSS = roundelNudges[roundelNudgeDirection];
    const hideRoundel = hideRoundelBelow ? roundelHidingPoints[hideRoundelBelow] : '';

    return (
        <div css={[hero, cssOverrides]}>
            {roundelText && (
                <div css={heroRoundelContainer}>
                    <div css={[heroRoundel, nudgeCSS, hideRoundel]}>{roundelText}</div>
                </div>
            )}
            <div css={useOffset ? roundelOffset : ''}>{children}</div>
            <div css={heroImage}>{image}</div>
        </div>
    );
}

Hero.defaultProps = {
    cssOverrides: '',
    roundelText: null,
    roundelNudgeDirection: 'up',
    hideRoundelBelow: '',
};

export default Hero;
