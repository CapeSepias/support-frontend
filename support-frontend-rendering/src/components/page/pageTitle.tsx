/** @jsx jsx */
import { jsx, css, SerializedStyles } from '@emotion/react';
import { from } from '@guardian/src-foundations/mq';
import { brand, brandAlt, neutral } from '@guardian/src-foundations/palette';
import { space } from '@guardian/src-foundations';
import { titlepiece } from '@guardian/src-foundations/typography';
import { Container } from '@guardian/src-layout';

const guardianWeeklyBlue = '#66c2e9';
const paperSubscriptionsBlue = '#90DCFF';

type ThemeType = 'showcase' | 'digital' | 'weekly' | 'paper';

type PropTypes = {
    title: string;
    theme: ThemeType;
    cssOverrides?: string;
    children: React.ReactChild;
};

const themeColors: { [key in ThemeType]: string } = {
    weekly: guardianWeeklyBlue,
    digital: brand[300],
    showcase: brandAlt[400],
    paper: paperSubscriptionsBlue,
};

const headerThemes: { [key in ThemeType]: SerializedStyles } = {
    weekly: css`
        :before {
            background-color: ${themeColors.weekly};
        }
    `,
    digital: css`
        color: ${neutral[97]};
        :before {
            background-color: ${themeColors.digital};
        }
    `,
    showcase: css`
        :before {
            background-color: ${themeColors.showcase};
        }
    `,
    paper: css`
        :before {
            background-color: ${themeColors.paper};
        }
    `,
};

const header = css`
    color: ${neutral[7]};
    position: relative;
    background-color: ${neutral[93]};
    display: flex;
    flex-direction: column;

    :before {
        position: absolute;
        top: 0;
        left: 0;
        width: 100%;
        height: 100%;
        content: '';
    }

    ${from.desktop} {
        :before {
            height: 370px;
        }
    }
`;

export const pageTitle = css`
    position: relative;
    ${titlepiece.small({ fontWeight: 'bold' })};
    z-index: 10;
    padding: ${space[9]}px ${space[4]}px;
    width: 100%;

    ${from.phablet} {
        width: 100%;
        align-self: center;
    }

    ${from.desktop} {
        ${titlepiece.large({ fontWeight: 'bold' })}
        margin: 0 auto;
        max-width: 1290px;
    }
`;

function PageTitle({ title, theme, cssOverrides, children }: PropTypes): React.ReactElement {
    return (
        <div css={[header, headerThemes[theme], cssOverrides]}>
            <Container>
                <h1 css={pageTitle}>{title}</h1>
            </Container>
            {children}
        </div>
    );
}

PageTitle.defaultProps = {
    cssOverrides: '',
};

export default PageTitle;
