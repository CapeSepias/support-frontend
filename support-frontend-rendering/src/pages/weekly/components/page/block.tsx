/** @jsx jsx */
import { jsx, css } from '@emotion/react';
import { neutral } from '@guardian/src-foundations/palette';
import { space } from '@guardian/src-foundations';
import { from } from '@guardian/src-foundations/mq';

type PropTypes = {
    children: React.ReactChild;
};

const block = css`
    position: relative;
    margin: ${space[6]}px 0;
    padding: ${space[9]}px 0;
    border: 1px solid ${neutral[86]};
    background-color: ${neutral[100]};
    z-index: 2;

    ${from.desktop} {
        padding: ${space[12]}px;
    }
`;

function Block(props: PropTypes): React.ReactElement {
    return <div css={block}>{props.children}</div>;
}

export default Block;
