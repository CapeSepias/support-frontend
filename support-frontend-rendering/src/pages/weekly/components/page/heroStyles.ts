import { css, SerializedStyles } from '@emotion/react';
import { brand, brandAlt, neutral } from '@guardian/src-foundations/palette';
import { Breakpoint, from, until } from '@guardian/src-foundations/mq';
import { space } from '@guardian/src-foundations';
import { headline, body } from '@guardian/src-foundations/typography';

const roundelSizeMob = 120;
const roundelSize = 180;

export const hero = css`
    position: relative;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    color: ${neutral[100]};
    border: none;
    padding-top: ${space[3]}px;
    background-color: ${brand[400]};
    width: 100%;

    ${from.tablet} {
        flex-direction: row;
    }

    /* Typography defaults */
    ${body.small()}

    ${from.mobileMedium} {
        ${body.medium()};
    }

    ${from.desktop} {
        ${headline.xxsmall()};
        line-height: 135%;
    }
    /* TODO: fix this when we port over the image components */
    .component-grid-picture {
        display: flex;
    }
`;

// On mobile the roundel can overlay and hide the h2 inside the hero
// This adds a little extra top margin if the roundel is present to keep the headline visible
export const roundelOffset = css`
    ${until.tablet} {
        margin-top: ${roundelSizeMob / 2 - space[6]}px;
    }
`;

export const heroImage = css`
    align-self: flex-end;
    flex-shrink: 0;
    display: flex;
    justify-content: flex-end;
    align-items: flex-end;
    width: 100%;

    ${from.tablet} {
        width: 40%;
    }

    & img {
        max-width: 100%;
    }
`;

export const heroRoundelContainer = css`
    position: absolute;
    top: 0;
    right: ${space[3]}px;

    ${from.tablet} {
        right: ${space[12]}px;
    }
`;

export const heroRoundel = css`
    display: flex;
    align-items: center;
    justify-content: center;
    float: right;
    text-align: center;
    transform: translateY(-67%);
    min-width: ${roundelSizeMob}px;
    max-width: ${roundelSize}px;
    width: calc(100% + ${space[3]}px);
    padding: ${space[1]}px;
    border-radius: 50%;
    background-color: ${brandAlt[400]};
    color: ${neutral[7]};
    ${headline.xxsmall({ fontWeight: 'bold' })};
    z-index: 20;

    ${from.tablet} {
        width: calc(100% + ${space[6]}px);
        transform: translateY(-50%);
        ${headline.small({ fontWeight: 'bold' })};
    }

    /* Combined with float on the main element, this makes the height match the content width for a perfect circle
  cf. https://medium.com/@kz228747/height-equals-width-pure-css-1c79794e470c */
    &::before {
        content: '';
        margin-top: 100%;
    }
`;

export const roundelNudgeUp = css`
    ${until.tablet} {
        transform: translateY(-67%);
    }
`;

export const roundelNudgeDown = css`
    ${until.tablet} {
        transform: translateY(-34%);
    }
`;

function hideUntilBreakpoint(breakpoint: Breakpoint): SerializedStyles {
    return css`
        ${until[breakpoint]} {
            display: none;
        }
    `;
}

export const roundelHidingPoints: { [key in Breakpoint]: SerializedStyles } = {
    mobile: hideUntilBreakpoint('mobile'),
    mobileMedium: hideUntilBreakpoint('mobileMedium'),
    mobileLandscape: hideUntilBreakpoint('mobileLandscape'),
    phablet: hideUntilBreakpoint('phablet'),
    tablet: hideUntilBreakpoint('tablet'),
    desktop: hideUntilBreakpoint('desktop'),
    leftCol: hideUntilBreakpoint('leftCol'),
    wide: hideUntilBreakpoint('wide'),
};
