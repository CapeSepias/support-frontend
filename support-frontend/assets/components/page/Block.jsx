// @flow

import React, { type Node } from 'react';
import { css } from '@emotion/core';
import { neutral } from '@guardian/src-foundations/palette';
import { space } from '@guardian/src-foundations';

type PropTypes = {|
  children: Node;
|};

const block = css`
  position: relative;
  margin: ${space[6]}px 0;
  border: 1px solid ${neutral[86]};
  background-color: ${neutral[100]};
  padding: ${space[12]}px;
  z-index: 2;
`;

function Block(props: PropTypes) {
  return (
    <div css={block}>
      {props.children}
    </div>
  );
}

export default Block;
