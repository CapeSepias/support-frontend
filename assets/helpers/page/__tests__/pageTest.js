// @flow

// ----- Imports ----- //

import type { Country } from 'helpers/internationalisation/country';

import { createCommonReducer } from '../page';


// ----- Tests ----- //

jest.mock('ophan', () => {});

describe('reducer tests', () => {

  let reducer = () => {};

  beforeEach(() => {

    const initialState = {
      campaign: 'dummy_campaign',
      acquisition: {
        referrerPageViewId: null,
        campaignCode: null,
        referrerUrl: null,
        componentId: null,
        componentTypeV2: null,
        source: null,
        abTests: [],
      },
      country: 'GB',
      abParticipations: {},
    };

    reducer = createCommonReducer(initialState);

  });

  it('should return the initial state', () => {
    expect(reducer(undefined, {})).toMatchSnapshot();
  });

  it('should handle SET_COUNTRY to US', () => {

    const country: Country = 'US';
    const action = {
      type: 'SET_COUNTRY',
      country,
    };

    const newState = reducer(undefined, action);

    expect(newState.country).toEqual(country);
    expect(newState.acquisition.campaignCode).toMatchSnapshot();
    expect(newState.campaign).toMatchSnapshot();
    expect(newState.acquisition.campaignCode).toMatchSnapshot();
    expect(newState.abParticipations).toMatchSnapshot();
  });

});
