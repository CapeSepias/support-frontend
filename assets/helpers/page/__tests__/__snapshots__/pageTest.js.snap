// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer tests should handle SET_COUNTRY to US 1`] = `null`;

exports[`reducer tests should handle SET_COUNTRY to US 2`] = `"dummy_campaign"`;

exports[`reducer tests should handle SET_COUNTRY to US 3`] = `null`;

exports[`reducer tests should handle SET_COUNTRY to US 4`] = `Object {}`;

exports[`reducer tests should return the initial state 1`] = `
Object {
  "abParticipations": Object {},
  "campaign": "dummy_campaign",
  "internationalisation": Object {
    "countryId": "GB",
  },
  "referrerAcquisitionData": Object {
    "abTests": Array [],
    "campaignCode": null,
    "componentId": null,
    "componentType": null,
    "referrerPageviewId": null,
    "referrerUrl": null,
    "source": null,
  },
}
`;
